packages/server/api/src/app/ai/chat/ai-tool-chat.controller.ts

```diff
@@ -0,0 +1,120 @@
+import { getAiProviderLanguageModel } from '@openops/common';
+import { logger } from '@openops/server-shared';
+import { PrincipalType } from '@openops/shared';
+import { Type } from '@sinclair/typebox';
+import { generateText, pipeDataStreamToResponse, streamText } from 'ai';
+import { encryptUtils } from '../../helper/encryption';
+import { aiConfigService } from '../config/ai-config.service';
+import { callMcpTool } from '../mcp-client/mcp-api';
+import { mcpTools } from '../mcp-client/mcp-tools';
+import { getChatHistory, saveChatHistory } from './ai-chat.service';
+
+const PersistentChatRequest = Type.Object({
+  chatId: Type.String(),
+  message: Type.String(),
+  model: Type.Optional(Type.String()),
+  system: Type.Optional(Type.String()),
+});
+
+const PersistentChatOptions = {
+  config: {
+    allowedPrincipals: [PrincipalType.USER],
+  },
+  schema: {
+    tags: ['ai', 'ai-chat'],
+    description:
+      'Persistent chat endpoint with tool calling support. LLM chooses and invokes MCP tools dynamically.',
+    body: PersistentChatRequest,
+    response: {
+      200: Type.Any(),
+    },
+  },
+};
+
+export const aiToolChatController: FastifyPluginAsyncTypebox = async (app) => {
+  app.post('/', PersistentChatOptions, async (request, reply) => {
+    const { chatId, message, model, system } = request.body;
+    const projectId = request.principal.projectId;
+
+    if (!chatId || !message) {
+      return reply.code(400).send({ error: 'Missing chatId or message.' });
+    }
+
+    const messages = await getChatHistory(chatId);
+    messages.push({ role: 'user', content: message });
+
+    try {
+      const aiConfig = await aiConfigService.getActiveConfigWithApiKey(
+        projectId,
+      );
+      if (!aiConfig) {
+        return await reply
+          .code(404)
+          .send('No active AI configuration found for the project.');
+      }
+      const languageModel = await getAiProviderLanguageModel({
+        apiKey: encryptUtils.decryptString(JSON.parse(aiConfig.apiKey)),
+        model: model || aiConfig.model,
+        provider: aiConfig.provider,
+        providerSettings: aiConfig.providerSettings,
+      });
+
+      const toolChoiceResult = await generateText({
+        model: languageModel,
+        messages,
+        tools: mcpTools,
+        toolChoice: 'auto',
+        ...aiConfig.modelSettings,
+      });
+
+      const toolCalls = toolChoiceResult.toolCalls;
+      logger.info(toolCalls, `toolCalls`);
+      if (toolCalls && toolCalls.length > 0) {
+        const { toolName, args: parameters } = toolCalls[0];
+
+        try {
+          logger.info(`toolName ${toolName}`);
+          logger.info(parameters, `parameters`);
+          const toolResult = await callMcpTool(toolName, parameters);
+          logger.info(toolResult, `toolResult`);
+
+          messages.push({
+            role: 'system',
+            content: `Tool (${toolName}) result: ${JSON.stringify(toolResult)}`,
+          });
```

Comment:
need to filter for the the UI tool messages
![image](https://github.com/user-attachments/assets/f791ed9a-6839-436a-a647-a2cda198e8f9)